oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,486 @@
+
+We also need a process to migrate core repositories through breaking changes in third-party dependencies. Ideally follow the `Upgrade Project Runbook <https://openedx.atlassian.net/wiki/spaces/AC/pages/3660316693/Upgrade+Project+Runbook>`_.
+
+Best Practices
+^^^^^^^^^^^^^^
+
+We need to ensure we minimize breaking changes in our own libraries (such as Paragon, the header, footer, frontend-platform, frontend-build, etc.) We suggest accomplishing this by:
+
+- Creating new versions of components with breaking changes (``ButtonV2``, ``webpack.dev.config.v2.js``) rather than modifying existing ones.
+- Leveraging the DEPR process for communication and removing old component versions
+- Aligning that removal and the subsequent breaking changes with Open edX releases, and documenting it in their release notes.
+
+Further, we could reduce the overhead of dependency maintenance and ensure MFEs stay up to date by pinning dependencies to major versions using ``^`` on versions in our package.json files.
+
+Tooling
+^^^^^^^
+
+Maintainers and developers should be warned of incompatibilities created by their PRs, or outside the repository by another project (such as the shell).
+
+This could take the form of Github tooling which notifies maintainers and developers that their frontend code has:
+
+1. Drifted behind the compatible version of a shared dependency for a given Open edX release or the main branch.
+2. Has upgraded beyond what is compatible with a given Open edX release or the main branch.
+
+Code Organization
+^^^^^^^^^^^^^^^^^
+
+We propose creating a passthrough library of shared dependencies. MFEs would depend on this single library in its package.json rather than on individual shared dependencies. This library would be versioned in accordance with the `Best Practices`_ suggestions above, meaning that breaking changes would be minimized and dependencies would be pinned to major versions via ``^`` on version numbers. Its version manifest would be the source of the version numbers for the `Process`_ and `Tooling`_ suggestions.
+
+To minimize impact on our MFEs, this library may need to be supported with some build-time configuration in Webpack that aliases its passthrough imports to their original package names. This would let us continue to write:
+
+  ``import React from 'react';``
+
+Instead of having to write something like:
+
+  ``import React from '@openedx/<passthrough library name>/react';``
+
+or:
+
+  ``import { React } from '@openedx/<passthrough library name>';``
+
+Out of Scope
+============
+
+There are a few important - but tangental - concerns which are considered out of scope for this OEP and its resulting reference implementation.
```

Comment:
SSR sounds like a nice optimization but from personal experience I actually prefer SPAs - the architecture is much simpler. SSR usually requires an actual web application to serve your frontend, as opposed to just dumping static files onto a CDN. And more annoyingly, "React Server Components" and "React Components" are very similar and often inter-mixed, but they behave differently and have different APIs available for sharing context, loading data, etc., and this makes the whole codebase much more complex - little gotchas everywhere.

Reply:
I think if we had had Node.JS as server instead of Django this path might easier, though there is a library https://github.com/markfinger/python-react that lets you render react on python based app/server, I doubt that is production ready anyway. 

I don't want to divert from the orignal topic, but still I feel very tempted to say, in the long run I think it's worth considering... Imagine a world where the unit/xblocks instead of being  _rendered through django-template/theme in learning MFE as iframe_ to be rendered simliar to other MFE, and one way that might happen if we support SSR in our current/future MFE 's ecosystem.  

Note: _I personally have strong opinion about the UX/UI of the learning MFE, because I believe it's the most important part of the product; a nice looking react compoenent would motivate learner to keep learning though a self-motivated learner would stick in any case :)_

Reply:
> SSR sounds like a nice optimization but from personal experience I actually prefer SPAs - the architecture is much simpler. SSR usually requires an actual web application to serve your frontend, as opposed to just dumping static files onto a CDN.

I agree SPAs are simpler architecturally, but I feel we shouldn't purely be optimizing for ease-of-development either; the rationale for SSR is more focused on users, and specifically the UX and performance (it's a _significant_ performance bottleneck for users to download JS/CSS assets at runtime today). As alluded to above, there's a reason React [officially recommends](https://react.dev/learn/start-a-new-react-project#can-i-use-react-without-a-framework) using a framework as a best practice at this point, and deprecated its `create-react-app` CLI to bootstrap new SPAs in favor of SSR frameworks.

> SSR usually requires an actual web application to serve your frontend.

Yes, it would introduce some infrastructure complexity on how to properly host SSR-compatible MFEs. IMHO, this seems like a fair tradeoff, though, to ensure Open edX MFEs are actually performant for users, especially on slower network speeds, where they are not today (e.g., there are **no** 2U/edX.org hosted MFEs with a Largest Contentful Paint in the "Good" threshold, largely due to the lengthy initial page load, and how most MFEs only start fetching data _after_ components have rendered, resulting in request waterfalls that are otherwise difficult to wrangle).

Code splitting and shared dependencies across MFEs can only get you so much in terms of frontend performance for standard/vanilla SPAs. For example, anecdotally, we just wrapped up a decently sized re-architecture of the `frontend-app-learner-portal-enterprise` MFE to migrate all its routes (~15-20 page routes) and API calls to React Router's route [loaders](https://reactrouter.com/en/main/route/loader) and React Query (for caching / automatic query retries); at this point we've largely exhausted most readily available code splitting options and other vanilla SPA "best practices" around performance (e.g., "Fetch-_Then_-Render" with route loaders, preconnect origins, etc.) yet had _little to no_ meaningful change in initial page load or Core Web Vitals (e.g., Largest Contentful Paint), despite the "Route change" duration decreasing significantly from ~15s to 5s (95th percentile); I feel this poor initial page load for users and seemingly inability to address it can be directly attributed to the lack of SSR support.

> I think if we had had Node.JS as server instead of Django this path might easier, though there is a library https://github.com/markfinger/python-react that lets you render react on python based app/server, I doubt that is production ready anyway.

When mentioning SSR as it relates to React, the assumption is generally that it'd be a Node.js server used. The SSR frameworks referred to here (e.g., Next.js, Remix, etc.) are specifically designed to work with React and all its nuances; Python-based services rendering standard React components are not.

FWIW, I believe `python-react` is fairly similar to the existing approach available in Django (at least in edx-platform) today to render React components ([source](https://github.com/openedx/edx-platform/blob/9d29595602e62c41d9b35731a8686d1c1d5e3609/common/djangoapps/pipeline_mako/templates/static_content.html#L158-L183)). Personally, I wouldn't necessarily consider a Python-based approach to be synonymous with modern SSR for React applications (e.g., using Remix).

Reply:
>  FWIW, I believe python-react is fairly similar to the existing approach available in Django (at least in edx-platform) today to render React components ([source](https://github.com/openedx/edx-platform/blob/9d29595602e62c41d9b35731a8686d1c1d5e3609/common/djangoapps/pipeline_mako/templates/static_content.html#L158-L183)). Personally, I wouldn't necessarily consider a Python-based approach to be synonymous with modern SSR for React applications (e.g., using Remix). 

I didn't know about edx-platform rendering of react, it seems it relies on this django https://github.com/django-webpack/django-webpack-loader app... 
I need to dig into it to know how it works, or/and compare to Remix... nonetheless I am not implying that the current defacto python based react render  (be it `django-webpack-loader` or `python-react`)is compatible with modern React SSR, rather to show it might be possible or as a building block. 

my conclusion if we are on the same page;  to be able to support the modern SSR either 1) develop python based service/app that support modern SSR or 2) we need to depend on Node.js type of service next, remix or alike. 

Lastly what I envision is, the SSR part react would be complitable with MFE, such that it utilize  [hydrateRoot](https://react.dev/reference/react-dom/client/hydrateRoot)

Reply:
@adamstankiewicz You make good points. Though if we're going to fix that by adopting SSR, then I think we might as well adopt either Next.js or Remix to take advantage of everything that a framework provides. At least with a framework there are conventions, tools, documentation, and fellow travellers to make dealing with the complexities of SSR much more manageable.

(I still am of the opinion that the way that server components and client components get smudged together creates an annoying DX, but I realize there are trade-offs here...)

Reply:
> Yes, it would introduce some infrastructure complexity on how to properly host SSR-compatible MFEs. IMHO, this seems like a fair tradeoff, though, to ensure Open edX MFEs are actually performant for users, especially on slower network speeds, 

One thing I do want to gently point out is that not all operators are OpenCraft or 2U or eduNEXT. I don't actually know the answer to this question: to what extent do we, as the project, want to encourage the ability for a small operator, perhaps a single person, to be able to stand up an instance without a great depth of DevOps knowledge? What does a better job of increasing access to knowledge, the very optimal user experience or a less optimal one that their rural high school admin can set up? This is a conversation I have idly had with Jenna, with no answer.

Reply:
To answer Adam's original question: no, SSR is not being considered here.  And neither is a framework.  The whole point of picking vanilla Webpack Module Federation over Piral was so we wouldn't have to refactor too much.  It sounds to me like Next.js would be much worse in that respect.

Are we prioritizing developers and operators over the users, then?  No.  It's just that moving from SPA to SSR is a significantly bigger lift than adopting Webpack Module federation plus some tooling and conventions around dependencies, all of which bring their own benefits.  Not only that, Module Federation is not the only thing that might preclude the adoption of a frontend framework - as we've been saying, it's a whole other ballpark.  As such, we should consider it separately.

(And since this has been brought up in this thread, for what it's worth I think we should totally be doing some refactoring around route loaders and React query - even if for nothing else than how much more sane it makes the architecture.  But, again, that should be a separate OEP/ADR.)

Reply:
> consider whether Server-Side Rendering (SSR) is compatible with composable MFEs with module federation to not preclude that from being an option in the future.

SSR with MF examples exist here https://github.com/module-federation/module-federation-examples/tree/master/server-side-rendering so I don't think this work will *prevent* any future SSR work.